exercise/static/exercise/duplicate_check.js

```diff
@@ -0,0 +1,67 @@
+function openDuplicateModalOrSubmit(exercise, hashes, hash, submitCallback) {
+  const isDuplicate = hashes.includes(hash);
+  if (isDuplicate) {
+    // Set the number of the existing duplicate submission
+    $("#duplicate-submission-modal").find("[data-dup-submission]").text(hashes.reverse().indexOf(hash) + 1);
+    // Unbind previous event handlers and bind a new one
+    $("#duplicate-submission-modal-button").unbind('click');
+    $("#duplicate-submission-modal-button").click(function() {
+      submitCallback(exercise, hash);
+    });
+    $("#duplicate-submission-modal").modal()
+  } else {
+    submitCallback(exercise, hash);
+  }
+};
+
+function duplicateCheck(exercise, form_element, submitCallback) {
+  // Retrieve all the hashes of the previous submissions from the HTML
+  const hashes = [];
+  const datahashElements = exercise.element.find("[data-hash]");
+  for (let elem of datahashElements) {
+    try {
+      hashes.push(elem.dataset["hash"]);
+    } catch(error) {
+    }
+  }
+
+  // Compute a hash for the current submission
+  const formData = new FormData(form_element);
+  const formProps = Object.fromEntries(formData);
+  // Grader language shouldn't affect the hash
+  if (formProps.hasOwnProperty('__grader_lang')) {
+    delete formProps['__grader_lang'];
+  }
+  let hashThis = JSON.stringify(formProps, Object.keys(formProps).sort());
+
+  // File contents are not included in formProps, so we read the files separately
+  const inputFileElements = form_element.querySelectorAll('input[type=file]');
+  if (inputFileElements.length > 0) {
```

Comment:
If the same form contains both file upload fields and other form fields like text inputs and radio buttons, then does this code compute the md5 hash only from the files or from all inputs? I just want to verify it. I get the impression that all fields are included and that is the goal.

Reply:
Yes, all form fields are added to the variable ``hashThis`` first, and if there are file fields present, the files are read and concatenated with the JSON string of the other fields.